lib/hmr/HotModuleReplacement.runtime.js

```diff
@@ -312,58 +317,57 @@ module.exports = function () {
 			.filter(Boolean);
 
 		if (errors.length > 0) {
-			setStatus("abort");
-			return Promise.resolve().then(function () {
+			return setStatus("abort").then(function () {
 				throw errors[0];
 			});
 		}
 
 		// Now in "dispose" phase
-		setStatus("dispose");
-
-		results.forEach(function (result) {
-			if (result.dispose) result.dispose();
-		});
-
-		// Now in "apply" phase
-		setStatus("apply");
-
-		var error;
-		var reportError = function (err) {
-			if (!error) error = err;
-		};
+		return setStatus("dispose").then(function () {
+			results.forEach(function (result) {
+				if (result.dispose) result.dispose();
+			});
 
-		var outdatedModules = [];
-		results.forEach(function (result) {
-			if (result.apply) {
-				var modules = result.apply(reportError);
-				if (modules) {
-					for (var i = 0; i < modules.length; i++) {
-						outdatedModules.push(modules[i]);
+			// Now in "apply" phase
+			return setStatus("apply").then(function () {
```

Comment:
Here is the only problem I'm seeing.

By this being a Promise, dispose and apply does no longer run in a single go.
This introduces the ability that user code is run in between, while the runtime is in a incomplete state.

e. g. A timeout could fire which imports a module, which is correctly in process of being replaced.

We can't introduce a break here, but I think it would work to all `setState("apply")` and await the Promise later (before `setStatus("idle")`/`setStatus("fail")`).

Reply:
That will complicate documentation somewhat, but I agree that it's the best option. I considered just not supporting `Promise`s at all for `dispose` and `apply`, but it seemed uglier to support `Promise`s for some statuses but not others, and this does provide the somewhat useful ability to write code that always runs after apply is finished regardless of whether there is an error or not.

Reply:
You can explain it as "running in parallel to the normal apply logic".